src/osd.c

```diff
@@ -458,13 +455,24 @@ display_osd(struct output *output)
 	cursor_update_focus(server);
 }
 
+static int
+nr_entries(struct server *server)
+{
+	struct wl_array views;
+	wl_array_init(&views);
+	view_array_append(server, &views,
+		LAB_VIEW_CRITERIA_CURRENT_WORKSPACE
+		| LAB_VIEW_CRITERIA_ROOT_TOPLEVEL
+		| LAB_VIEW_CRITERIA_NO_SKIP_WINDOW_SWITCHER);
+	int count = wl_array_len(&views);
+	wl_array_release(&views);
+	return count;
+}
+
 void
 osd_update(struct server *server)
 {
-	struct wl_list *node_list =
-		&server->workspace_current->tree->children;
-
-	if (wl_list_empty(node_list) || !server->osd_state.cycle_view) {
+	if (!nr_entries(server) || !server->osd_state.cycle_view) {
```

Comment:
We are now creating the same array at least two and at most 1 + output count times so I'd suggest creating the array once in `osd_update` and then pass it around to `display_osd()` instead.